core/store/src/lib.rs

```diff
@@ -284,27 +309,77 @@ impl NodeStorage {
 }
 
 impl Store {
-    pub fn new(storage: Arc<dyn Database>) -> Self {
-        Self { storage }
+    pub fn new(storage: Arc<dyn Database>, shard_uid_mapping: ShardUIdMapping) -> Self {
+        Self { storage, shard_uid_mapping }
+    }
+
+    /// Underlying `get()` implementation for all columns.
+    fn get_impl(&self, column: DBCol, key: &[u8]) -> io::Result<Option<DBSlice<'_>>> {
+        let value = if column.is_rc() {
+            self.storage.get_with_rc_stripped(column, &key)
+        } else {
+            self.storage.get_raw_bytes(column, &key)
+        }?;
+        Ok(value)
+    }
+
+    /// Reads shard_uid mapping for given shard and updates the in-memory mapping.
+    fn read_shard_uid_mapping_from_db(&self, shard_uid: ShardUId) -> io::Result<ShardUId> {
+        let mapped_shard_uid =
+            self.get_ser::<ShardUId>(DBCol::ShardUIdMapping, &shard_uid.to_bytes())?;
+        Ok(self.shard_uid_mapping.update(&shard_uid, mapped_shard_uid))
+    }
+
+    /// Specialized `get` implementation for State column that replaces shard_uid prefix
+    /// with a mapped value according to mapping strategy in Resharding V3.
+    ///
+    /// It first uses the in-memory `shard_uid_mapping` and attempts to read from storage.
+    /// If the node was not found in storage, it tries to update the in-memory mapping with
+    /// the mapping stored in `DBCol::ShardUIdMapping`.
+    /// If it was different, then we do the second attempt to read from storage.
+    fn get_impl_state(&self, key: &[u8]) -> io::Result<Option<DBSlice<'_>>> {
```

Comment:
Umm, why do we need this logic to be so complicated? Can't we just keep the singleton instance of mapping and DBCol::ShardUIdMapping in sync? Update one with the other?

Reply:
Also, should be quite cheap to read from store to build mapping once as long as it's not happening too frequently.

Reply:
> Also, should be quite cheap to read from store to build mapping once as long as it's not happening too frequently.

That would require retrieving all shard uids and passing it to the constructor of the mapping. Then we potentially fail to read from db in the constructor so it should return an error, and it would require changing signature of the places where we construct the mapping. For simplicity, I've chosen not to do this.

> Umm, why do we need this logic to be so complicated? Can't we just keep the singleton instance of mapping and DBCol::ShardUIdMapping in sync? Update one with the other?

Updating db mapping based on in-mem mapping is problematic, as it requires in-mem mapping to know resharding history. I've chosen db mapping as the source of truth, and make in-mem mapping synchronize to db mapping.

I was thinking about how do we keep these two in sync and considered these two approaches:
- we call `in_mem_mapping.update()` when we update the mapping stored in db
- we potentially sync to db when we use in_mem_mapping

I used the second approach, as it only requires a few lines of code in `get_impl_state` and does not require the extra logic from the first approach. The extra logic would be to make sure we update all instances of `ShardUIdMapping`. It acts like a singleton, but beside neard client we have many tools that also construct it, so it would be quite a lot of changes. Also, we do not have yet the logic that updates the in-db mapping, e.g. on resharding event, so I would have to create a stub for it. But when we have it, we can reconsider simplifying `get_impl_state` and using the first approach, wdyt?